File: Civitas_GUI/src/civitas/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;
import java.util.ArrayList;
import GUI.Dado;

public class Jugador implements Comparable<Jugador> {
    
    protected static final int CasasMax = 4;
    protected static final int CasasPorHotel = 4;
    protected static final int HotelesMax = 4;
    protected static final float PasoPorSalida = 1000;
    protected static final float PrecioLibertad = 200;
    private static final float SaldoInicial = 7500;
    protected boolean encarcelado;
    protected String nombre;
    protected int numCasillaActual;
    protected boolean puedeComprar;
    protected float saldo;
    protected ArrayList<TituloPropiedad> propiedades;
    protected Sorpresa salvoconducto;
    
    Jugador(String name){
        nombre = name;
        propiedades = new ArrayList<>();
        encarcelado = false;
        numCasillaActual = 0;
        puedeComprar = true;
        saldo = SaldoInicial;
        salvoconducto = null;
    }
    
    protected Jugador(Jugador orig){
        nombre = orig.nombre;
        propiedades = new ArrayList<>();
        for(int i = 0; i < orig.propiedades.size(); i++)
            propiedades.add(orig.propiedades.get(i));
        encarcelado = orig.encarcelado;
        numCasillaActual = orig.numCasillaActual;
        puedeComprar = orig.puedeComprar;
        saldo = orig.saldo;
        salvoconducto = orig.salvoconducto;
    }
    
    public String getNombre(){
        return nombre;
    }
    
    int cantidadCasasHoteles(){
        int amount = 0;
        
        for(int i = 0; i < propiedades.size(); i++)
            amount += propiedades.get(i).cantidadCasasHoteles();
        
        return amount;
    }
    
    Boolean enBancarrota(){
        return saldo <= 0;
    }
    
    private Boolean existeLaPropiedad(int ip){
        return ip >= 0 && ip < propiedades.size();
    }
    
    private int getCasasMax(){
        return CasasMax;
    }
    
    int getCasasPorHotel(){
        return CasasPorHotel;
    }
    
    private int getHotelesMax(){
        return HotelesMax;
    }
    
    int getNumCasillaActual(){
        return numCasillaActual;
    }
    
    private float getPrecioLibertad(){
        return PrecioLibertad;
    }
    
    private float getPremioPasoSalida(){
        return PasoPorSalida;
    }
    
    public ArrayList<TituloPropiedad> getPropiedades(){
        return propiedades;
    }
    
    Boolean getPuedeComprar(){
        return puedeComprar;
    }
    
    public float getSaldo(){
        return saldo;
    }
    
    public Boolean isEncarcelado(){
        return encarcelado;
    }
    
    public void setPuedeComprar(boolean puede){
        puedeComprar = puede;
    }
    
    private Boolean puedoEdificarCasa(TituloPropiedad propiedad){
        return propiedad.getNumCasas() < CasasMax && saldo >= propiedad.getPrecioEdificar() && propiedades.contains(propiedad);
    }
    
    private Boolean puedoEdificarHotel(TituloPropiedad propiedad){
        return propiedad.getNumHoteles() < HotelesMax && saldo >= propiedad.getPrecioEdificar() && propiedades.contains(propiedad) && propiedad.getNumCasas() == CasasMax;
    }
    
    private Boolean debeSerEncarcelado(){
        if(encarcelado)
            return false;
        else{
            if(!tieneSalvoconducto())
                return true;
            else{
                perderSalvoconducto();
                Diario.getInstance().ocurreEvento("¡" + nombre + " se libra de la cárcel!");
                return false;
            }
        }
    }
    
    Boolean encarcelar(int numCasillaCarcel){
        if(debeSerEncarcelado()){
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            Diario.getInstance().ocurreEvento(nombre + " ha sido encarcelado");
        }
        
        return encarcelado;
    }
    
    Boolean obtenerSalvoconducto(Sorpresa s){
        if(encarcelado)
            return false;
        else{
            salvoconducto = s;
            return true;
        }
    }
    
    protected void perderSalvoconducto(){
        if(salvoconducto != null){
            ((Sorpresa_salirCarcel) salvoconducto).usada();
            salvoconducto = null;
        }
    }
    
    Boolean tieneSalvoconducto(){
        return salvoconducto != null;
    }
    
    Boolean puedeComprarCasilla(){
        puedeComprar = !encarcelado;
        
        return puedeComprar;
    }
    
    Boolean paga(float cantidad){
        return modificarSaldo(cantidad*(-1));
    }
    
    Boolean pagaImpuesto(float cantidad){
        if(encarcelado)
            return false;
        else
            return paga(cantidad);
    }
    
    Boolean pagaAlquiler(float cantidad){
        if(encarcelado)
            return false;
        else
            return paga(cantidad);
    }
    
    Boolean recibe(float cantidad){
        if(encarcelado)
            return false;
        else
            return modificarSaldo(cantidad);
    }
    
    Boolean modificarSaldo(float cantidad){
        saldo += cantidad;
        
        if(cantidad >= 0)
            Diario.getInstance().ocurreEvento("El saldo de " + nombre + " se ha incrementado en " + cantidad);
        else
            Diario.getInstance().ocurreEvento("El saldo de " + nombre + " se ha decrementado en " + (cantidad*-1));
        
        return true;
    }
    
    Boolean moverACasilla(int numCasilla){
        if(encarcelado)
            return false;
        else{
            numCasillaActual = numCasilla;
            //puedeComprar = false;
            Diario.getInstance().ocurreEvento(nombre + " se mueve a la casilla nº " + numCasilla);
            return true;
        }
    }
    
    private Boolean puedoGastar(float cantidad){
        if(encarcelado)
            return false;
        else
            return saldo >= cantidad;
    }
    
    Boolean vender(int ip){
        if(encarcelado)
            return false;
        else{
            if(existeLaPropiedad(ip)){ 
               if(propiedades.get(ip).vender(this)){
                   Diario.getInstance().ocurreEvento(nombre + " ha vendido " + propiedades.get(ip).getNombre());
                   propiedades.remove(ip);
                   return true;
               }
               
               else
                return false;
            }
            
            else
                return false;
        }
    }
    
    Boolean tieneAlgoQueGestionar(){
        return !propiedades.isEmpty();
    }
    
    protected Boolean puedeSalirCarcelPagando(){
        return saldo >= PrecioLibertad;
    }
    
    Boolean salirCarcelPagando(){
        if(encarcelado && puedeSalirCarcelPagando()){
            if(paga(PrecioLibertad)){
                encarcelado = false;
                Diario.getInstance().ocurreEvento(nombre + " ha comprado su libertad y ya no está encarcelado");
                return true;
            }
            
            else
                return false;
        }
        
        else
            return false;
    }
    
    Boolean salirCarcelTirando(){
        if(Dado.getInstance().salgoDeLaCarcel()){
            encarcelado = false;
            Diario.getInstance().ocurreEvento("¡" + nombre + " ha conseguido salir de la cárcel!");
            return true;
        }
        else
            return false;
    }
    
    Boolean pasaPorSalida(){
        modificarSaldo(PasoPorSalida);
        Diario.getInstance().ocurreEvento(nombre + " ha pasado por la salida y recibe su bonus de " + PasoPorSalida);
        return true;
    }
    
    @Override
    public int compareTo(Jugador otro) {
        return (new Float(this.getSaldo()).compareTo(otro.getSaldo()));
    }
    
    @Override
    public String toString(){
        String cadena = nombre + "\n Casilla Actual: " + numCasillaActual + "\n Saldo: " + saldo + "\n Propiedades: ";
        if(propiedades.isEmpty())
            cadena += "Ninguna";
        else{
            for(int i = 0; i < propiedades.size(); i++)
                cadena += propiedades.get(i).getNombre() + ", ";
        }
        
        cadena += "\n Puede Comprar: ";
        
        if(puedeComprar)
            cadena += "Si";
        else
            cadena += "No";
        
        cadena += "\n Está Encarcelado: ";
        
        if(encarcelado)
            cadena += "SI";
        else
            cadena += "NO";
        
        return cadena;
    }
    
    Boolean construirCasa(int ip){
         Boolean result = false;
         if(encarcelado)
             return result;
         
         if(!encarcelado && existeLaPropiedad(ip)){
             TituloPropiedad propiedad = propiedades.get(ip);
             Boolean puedoEdificarCasa = puedoEdificarCasa(propiedad);
             
             if(puedoEdificarCasa){
                 result = propiedad.construirCasa(this);
                 Diario.getInstance().ocurreEvento("El jugador " + nombre + " construye una casa en "+ propiedades.get(ip).getNombre());
             }
         }
         
         return result;
    }
    
    Boolean construirHotel(int ip){
        Boolean result = false;
        if(encarcelado)
            return result;
        
        if(existeLaPropiedad(ip)){
            TituloPropiedad propiedad = propiedades.get(ip);
            Boolean puedoEdificarHotel = puedoEdificarHotel(propiedad);            
            
            if (puedoEdificarHotel){
                result = propiedad.construirHotel(this);
                propiedad.derruirCasas(CasasPorHotel, this);
                Diario.getInstance().ocurreEvento("El jugador " +nombre+ " construye un hotel en " + propiedades.get(ip).getNombre());
            }                       
        }
        
        return result;
    }
    
    Boolean hipotecar(int ip){
        Boolean result = false;
        if (encarcelado)
           return result;
        
        if (existeLaPropiedad(ip)){
            TituloPropiedad propiedad = propiedades.get(ip);
            result = propiedad.hipotecar(this);
        }
        
        if (result)
            Diario.getInstance().ocurreEvento("El jugador " +nombre+ " hipoteca la propiedad "+ip);
        
        return result;
    }
    
    Boolean cancelarHipoteca(int ip){
        Boolean result = false;
        
        if (encarcelado)
            return result;
        
        if(this.existeLaPropiedad(ip)){
            
            TituloPropiedad propiedad = propiedades.get(ip);
            float cantidad = propiedad.getImporteCancelarHipoteca();
            Boolean puedoGastar = this.puedoGastar(cantidad);
            
            if(puedoGastar){
                result = propiedad.cancelarHipoteca(this);
                
                if(result)
                    Diario.getInstance().ocurreEvento("El jugador " + nombre + " cancela la hipoteca de la propiedad " + propiedad.getNombre());
            }
        }
  
        return result;
    }
    
    Boolean comprar(TituloPropiedad titulo){
        Boolean result = false;
        
        if(encarcelado)
            return result;
        
        if (puedeComprar){
            float precio = titulo.getPrecioCompra();
            
            if(puedoGastar(precio)){
                result = titulo.comprar(this);
            }
        }
        
        if (result){
            propiedades.add(titulo);
            Diario.getInstance().ocurreEvento("El jugador " + nombre + " compra la propiedad " + titulo.toString());
            System.out.println("Se ha efectuado la compra");
        }
        
        return result;
    }
    
}

File: Civitas_Ruby/lib/jugador.rb
#!/bin/env ruby
# encoding: utf-8
require_relative "titulo_propiedad.rb"
require_relative 'dado.rb'
require_relative 'diario.rb'

module Civitas
  class Jugador
    
    attr_reader :nombre, :numCasillaActual, :propiedades, :puedeComprar, :saldo , :encarcelado, :puedeComprar, :salvoconducto
    
    @@CASASMAX = 4
    @@CASASPORHOTEL = 4
    @@PASOPORSALIDA = 1000
    @@PRECIOLIBERTAD = 200
    @@HOTELESMAX = 4
    @@SALDOINICIAL = 7500
    
    def initialize(name)
      @nombre = name
      @propiedades = []
      @encarcelado = false
      @numCasillaActual = 0
      @puedeComprar = true
      @saldo = @@SALDOINICIAL
      @salvoconducto = nil
    end
    
    def self.copia(orig)
      jugador = new(orig.nombre)
      jugador.from(orig)
      return jugador
    end
    
    def from(orig)
      @nombre = orig.nombre
      @propiedades = []
      for i in (0..orig.propiedades.length-1)
        @propiedades.push(orig.propiedades[i])
      end
      @encarcelado = orig.encarcelado
      @numCasillaActual = orig.numCasillaActual
      @puedeComprar = orig.puedeComprar
      @saldo = orig.saldo
      @salvoconducto = orig.salvoconducto
    end
    
    def self.getCasasMax()
      return @@CASASMAX
    end
    
    def self.getCasasPorHotel()
      return @@CASASPORHOTEL
    end
    
    def self.getHotelesMax()
      return @@HOTELESMAX
    end
    
    def self.getPremioPasoPorSalida()
      return @@PASOPORSALIDA
    end
    
    def self.getPrecioLibertad
      return @@PRECIOLIBERTAD
    end
    
    def cantidadCasasHoteles()
      amount = 0
      
      for i in (0..@propiedades.length)
        amount += @propiedades[i].cantidadCasasHoteles()
      end
      
      return amount
    end
    
    def enBancarrota()
      return @saldo <= 0
    end
    
    def existeLaPropiedad(ip)
      return ip >= 0 && ip < @propiedades.length
    end
    
    def puedoEdificarCasa(propiedad)
      return propiedad.numCasas < @@CASASMAX && @saldo >= propiedad.precioEdificar && @propiedades.include?(propiedad)
    end
    
    def puedoEdificarHotel(propiedad)
      return propiedad.numHoteles < @@HOTELESMAX && @saldo >= propiedad.precioEdificar && @propiedades.include?(propiedad) && propiedad.numCasas == @@CASASMAX
    end
    
    def debeSerEncarcelado()
      if @encarcelado
        return false
      else
        if !tieneSalvoconducto()
          return true
         
        else
          perderSalvoConducto()
          Diario.instance.ocurre_evento("¡ #{@nombre} se libra de la cárcel!")
          return false
        end
        
      end
      
    end
    
    def encarcelar(numCasillaCarcel)
      if debeSerEncarcelado()
        moverACasilla(numCasillaCarcel)
        @encarcelado = true
        Diario.instance.ocurre_evento("#{@nombre} ha sido encarcelado")
      end
      
      @encarcelado
    end
    
    def obtenerSalvoconducto(sorpresa)
      if @encarcelado
        return false
      else
        @salvoconducto = sorpresa
        return true
        
      end
    end
    
    def perderSalvoConducto()
      if @salvoconducto != nil
        @salvoconducto.usada()
        @salvoconducto = nil
      end
    end
    
    def tieneSalvoconducto()
      return @salvoconducto != nil
    end
    
    def puedeComprarCasilla()
      @puedeComprar = !@encarcelado
      @puedeComprar
    end
    
    def paga(cantidad)
      return modificarSaldo(cantidad*(-1))
    end
    
    def pagaImpuesto(cantidad)
      if @encarcelado
        return false
      else
        return paga(cantidad)
      end
    end
    
    def pagaAlquiler(cantidad)
      if(@encarcelado)
        return false
      else
        return paga(cantidad)
      end
    end
    
    def recibe(cantidad)
      if @encarcelado
        return false
        
      else
        return modificarSaldo(cantidad)
      end
    end
    
    def modificarSaldo(cantidad)
      @saldo += cantidad
      
      if cantidad >= 0
        Diario.instance.ocurre_evento("El saldo de #{@nombre} se ha incrementado #{cantidad}")
      else
        Diario.instance.ocurre_evento("El saldo de #{@nombre} se ha decrementado #{cantidad}") 
      end
    end
    
    def moverACasilla(numCasilla)
      if @encarcelado
        return false
        
      else
        @numCasillaActual = numCasilla
        #@puedeComprar = false
        Diario.instance.ocurre_evento("#{@nombre} se me mueve a la casilla nº #{numCasilla}")
      end
    end
    
    def puedoGastar(cantidad)
      if @encarcelado
        return false
      else
        return @saldo >= cantidad
      end
    end
    
    def vender(ip)
      if @encarcelado
        return false
      else
        if(existeLaPropiedad(ip))
          if @propiedades[ip].vender(self)
            Diario.instance.ocurre_evento("#{@nombre} ha vendido #{@propiedades[ip].nombre}")
            @propiedades.delete_at(ip);
            return true
          else
            return false
          end
          
        else
          return false
        end
      end
    end
    
    def tieneAlgoQueGestionar()     
      return !@propiedades.empty?
    end
    
    def puedeSalirCarcelPagando()
      return @saldo >= @@PRECIOLIBERTAD
    end
    
    def salirCarcelPagando()
      if @encarcelado && puedeSalirCarcelPagando()
        if paga(@@PRECIOLIBERTAD)
          @encarcelado = false
          Diario.instance.ocurre_evento("#{@nombre} ha comprado su libertad y ya no esta encarcelado")
          return true
          
        else
          return false
        end
        
      else
        return false
      end
    end
    
    def salirCarcelTirando()
      
      puts "\n\n se esta repitiendo el salirCarceTirando y no deberia \n\n"
      
      if Dado.instance.salgoDeLaCarcel()
        @encarcelado = false
        Diario.instance.ocurre_evento("¡#{@nombre} ha conseguido salir de la cárcel!")
        return true
      else
        return false
      end
    end
    
    def pasaPorSalida()
      modificarSaldo(@@PASOPORSALIDA)
      Diario.instance.ocurre_evento("#{@nombr} ha pasado por la salida y recibe bonus de #{@@PASOPORSALIDA}")
      return true
    end
    
    def <=>(otro)
      return @saldo <=> otro.saldo
    end
    
    def to_s()
      cadena = "#{@nombre} \n Casilla Actual: #{@numCasillaActual} \nSaldo: #{@saldo} \nPropiedades: "
      if(!@propiedades.empty?)
        for i in (0..@propiedades.length-1)
          cadena += "#{@propiedades[i].nombre}, "
        end
      else
        cadena += "Ninguna"
      end
      
      cadena += "\n Puede Comprar: "
      
      if @puedeComprar
        cadena += "SI"
      else
        cadena += "NO"
      end
      
      cadena += "\n Está encarcelado: "
      
      if @encarcelado
        cadena += "SI"
      else
        cadena += "NO"
      end
      
      cadena
    end
    
    def construirCasa(ip)
      result = false
      if(@encarcelado)
        return result
      end
      
      if(!@encarcelado && existeLaPropiedad(ip))
        propiedad = @propiedades[ip]
        puedoEdificarCasa = puedoEdificarCasa(propiedad)
        
        if(puedoEdificarCasa)
          result = propiedad.construirCasa(self);
          Diario.instance.ocurre_evento("El jugador #{@nombre} construye casa en la propiedad #{ip}")
        end
        
      end
      
      return result
      
    end
    
    def construirHotel(ip)
      result = false
      if(@encarcelado)
        return result
      end
      
      if(!@encarcelado && existeLaPropiedad(ip))
        propiedad = @propiedades[ip]
        puedoEdificarHotel = puedoEdificarHotel(propiedad)
        
        if(puedoEdificarHotel)
          result = propiedad.construirHotel(self);
          propiedad.derruirCasas(@@CASASPORHOTEL, self)
          Diario.instance.ocurre_evento("El jugador #{@nombre} construye hotel en la propiedad #{ip}")
        end
        
      end
      
      return result
    end
    
    def hipotecar(ip)
      result = false
      
      if (@encarcelado)
        return result
      end
      
      if existeLaPropiedad(ip)
        propiedad = @propiedades[ip]
        result = propiedad.hipotecar(self)
      end
      
      if(result)
        Diario.instance.ocurre_evento("El jugador #{@nombre} hipoteca la propiedad #{ip}")
      end
      
      return result
    end
    
    def cancelarHipoteca(ip)
      result = false
      
      if(@encarcelado)
        return result
      end
      
      if (existeLaPropiedad(ip))
        titulo = @propiedades[ip]
        cantidad = titulo.getImporteCancelarHipoteca()
        puedoGastar = puedoGastar(cantidad)
        
        if puedoGastar
          result = titulo.cancelarHipoteca(self)
        end
      end
      if result
          Diario.instance.ocurre_evento("El jugador #{@nombre} cancela la hipoteca de #{ip}")
      end
      return result
    end
    
    def comprar(titulo)
      result = false
      
      if(@encarcelado)
        return result
      end
      
      if(@puedeComprar)
        precio = titulo.precioCompra()
        
        if(puedoGastar(precio))
          result = titulo.comprar(self)
        end
      end
      
      if(result)
        @propiedades.push(titulo)
        Diario.instance.ocurre_evento("El jugador #{@nombre} compra la propiedad #{titulo.to_s()}")
      end
    end
    
    def puede_comprar
      @puedeComprar
    end
    
    private :existeLaPropiedad, :perderSalvoConducto, :puedeSalirCarcelPagando, :puedoEdificarCasa, :puedoEdificarHotel, :puedoGastar
    protected :debeSerEncarcelado,  :saldo #:from #volver a colocar nombre :propiedades, poner a protected supuestamente
    private_class_method :getCasasMax, :getHotelesMax, :getPrecioLibertad, :getPremioPasoPorSalida
  end
end

